### PR TITLE
JP-3995: Make background step able to run with asn files as well

### PR DESCRIPTION
**Pull request by penaguerrero**
Resolves [JP-3995](https://jira.stsci.edu/browse/JP-3995)

This PR implements the option to run the step as a standalone with either fits or asn files as input. Also included, is the header keyword when the step completes or skips. This PR affects all modes but does not break any current pipeline. 

## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [ ] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by penaguerrero** [comment]
reg tests: https://github.com/spacetelescope/RegressionTests/actions/runs/15031186717

**Comment by penaguerrero** [comment]
clean regression test run! :)

**Comment by penaguerrero** [comment]
running reg tests: https://github.com/spacetelescope/RegressionTests/actions/runs/15329305686

**Review by melanieclarke — DISMISSED** [review]
Looks like there's still some more work to do here, to clean up the interfaces, code structure, and tests.

File: jwst/background/asn_intake.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                "Wrong number of science exposures found in {}".format(exp_product["name"])  # noqa: E501
+            )
+            log.warning("    Using only first one.")
+        science_member = science_member[0]
+        log.info("Working on input %s ...", science_member)
+
+        # Open the datamodel and update it with the relevant info for the background step
+        with datamodels.open(science_member) as sci:
+            exp_type = sci.meta.exposure.type
+            if exp_type in WFSS_TYPES:
```
[the remark this answers is not in the document]
Comment (penaguerrero) [reply jwst/background/asn_intake.py]: I tested removing the part that assigns the science, catalogue, and direct image in the datamodel assuming it is in the product of assign_wcs but it is not there, so the code crashes without that part.

[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/background/asn_intake.py]: Okay, I took a closer look at how the WFSS background step works and I think I understand it better now.

First, to get a product that includes the segmentation image, catalog, and direct image, save the assign_wcs product from a spec2 run.  For some NIRCam regtest data:
`strun calwebb_spec2 jw01076-o101_20220403t120233_spec2_002_asn.json --steps.assign_wcs.save_results=True`
The file names are saved in SCATFILE, SEGMFILE, and DIRIMAGE keys in the primary header.

From that product, no ASN is needed, and no background list either - just `strun background jw01076101001_02101_00003_nrcalong_assign_wcs.fits` should work.

But I can also see that it's reasonable someone might try:
```
strun assign_wcs jw01076101001_02101_00003_nrcalong_rate.fits
strun background jw01076-o101_asn.json
```
with the ASN file including the catalog and the assignwcsstep product.  I assume something like this is how you made the assign_wcs product you're testing with?

So I think you're right - it's okay to include here a check for the source catalog and image members in the ASN, if provided, and set the metadata appropriately if possible.

File: jwst/background/asn_intake.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                    log.info(f"Using segmentation map {sci.meta.segmentation_map}")
+                    sci.meta.direct_image = Path(members_by_type["direct_image"][0]).name
+                    log.info(f"Using direct image {sci.meta.direct_image}")
+                except IndexError:
+                    if sci.meta.source_catalog is None:
+                        raise IndexError(
+                            "No source catalog specified in association or datamodel."
+                        ) from None
+
+    return sci, members_by_type
```
Comment (melanieclarke) [inline jwst/background/asn_intake.py:70]: This return is outside the with block that opened `sci`, so it's returning a closed datamodel -- I think it should return the opened model. Also, it's looping over products, but only returning the last `sci` -- it should probably warn and take the first one instead.

File: jwst/background/background_step.py [hunk 3]
```diff
@@ -15,6 +18,7 @@ class BackgroundStep(Step):
     class_alias = "background"
 
     spec = """
+        bkg_list = force_list(default=None)  # List of background files
```
Comment (melanieclarke) [inline jwst/background/background_step.py]: Can you please add that this is ignored for WFSS or if an ASN is provided?

File: jwst/background/tests/test_background.py [hunk 4]
```diff
@@ -225,3 +226,80 @@ def test_miri_subarray_partial_overlap(data_shape, background_shape):
 
     image.close()
     background.close()
+
+
+def mk_asn(tmp_path):
```
Comment (melanieclarke) [inline jwst/background/tests/test_background.py]: This might be a little easier to use if you make it a fixture, and it returns the created filenames.  That way it could be called only once for the module, the files could be reused, and you wouldn't have to duplicate the filenames inside the test code.

File: jwst/background/tests/test_background.py [hunk 5]
```diff
@@ -225,3 +226,80 @@ def test_miri_subarray_partial_overlap(data_shape, background_shape):
...
+    image = miri_rate_model((100, 100), value=image_value)
+    image.save(str(tmp_path / "jw01000005001_test_mirimage_rate.fits"))
+    background = miri_rate_model((100, 100), value=background_value)
+    background.save(str(tmp_path / "jw01000005001_testbg_mirimage.fits"))
+    image.data = image.data - background.data
+    image.save(str(tmp_path / "jw010000-test_spec2_00001_asn_backgroundstep.fits"))
+    image.close()
+    background.close()
+
+
```
Comment (melanieclarke) [inline jwst/background/tests/test_background.py]: These tests look good - thanks for adding them!  It would also be helpful to have tests for the error conditions, with empty background lists.

File: jwst/background/tests/test_background_wfss.py
Comment (melanieclarke) [inline jwst/background/tests/test_background_wfss.py:1]: Can you add here an ASN-based test for WFSS input?

**Review by melanieclarke — COMMENTED** [review]
Getting closer, thanks for cleaning up the tests.

**Comment by penaguerrero** [comment]
new reg test: https://github.com/spacetelescope/RegressionTests/actions/runs/15475497992

File: jwst/background/background_step.py [hunk 6]
```diff
@@ -4,17 +4,21 @@
...
 
+
 __all__ = ["BackgroundStep"]
 
 
 class BackgroundStep(Step):
     """Subtract background exposures from target exposures."""
 
-    class_alias = "background"
+    class_alias = "back_sub"
```
Comment (melanieclarke) [inline jwst/background/background_step.py]: This should be "bkg_subtract" to match the pipeline alias, not "back_sub".

Comment (penaguerrero) [reply jwst/background/background_step.py]: I'm actually not convinced we can just switch from background... The change broke a bunch of tests... How about I revert it to background and we investigate that in another PR?

Comment (melanieclarke) [reply jwst/background/background_step.py]: The failures are just because "background" still exists in the suffix and config files -- those have to be changed if you change the step alias.  But let's go ahead and address it in a separate PR, no need to do it here.

Comment (penaguerrero) [reply jwst/background/background_step.py]: and by the way, it seems there is even more confusion than we thought, the datamodels repo has in the cal_steps section of the core schema "back_sub" as alias for the background step... definitely need to investigate all the different alias the step has to try to converge in one

**Comment by penaguerrero** [comment]
re-running reg tests: https://github.com/spacetelescope/RegressionTests/actions/runs/15498534379

File: jwst/background/background_step.py [hunk 7]
```diff
@@ -30,91 +34,114 @@ class BackgroundStep(Step):
...
+
+            # Make sure that the background list is not empty for this case,
+            # or report and skip the step
+            if bkg_list is None or len(bkg_list) == 0:
+                self.log.warning("* No background list provided * Skipping step.")
+                result.meta.cal_step.back_sub = "SKIPPED"
+                return result
+
+            # Make sure to catch a trailing comma and ignore it
+            bkg_list = [bg for bg in bkg_list if bg]
```
Comment (melanieclarke) [inline jwst/background/background_step.py]: This should be moved above the check for an empty list, just in case someone tries to send you just a comma.

File: jwst/background/background_step.py (hunk 6 above)
Comment (melanieclarke) [reply jwst/background/background_step.py]: Oh dear.  Yes, let's definitely defer handling this so we can sort out all the issues at once.

**Review by melanieclarke — COMMENTED** [review]
Okay, I think this looks good now, assuming nothing comes up in the regression tests.  I have one small suggestion for the trailing comma check.

**Comment by penaguerrero** [comment]
new reg test run all clean: https://github.com/spacetelescope/RegressionTests/actions/runs/15540939358
